src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,386 @@
+    /// Variable-base scalar multiplication (lo half)
+    pub q_mul_lo: Selector,
+    /// Selector used in scalar decomposition for variable-base scalar mul
+    pub q_mul_decompose_var: Selector,
+    /// Variable-base scalar multiplication (final scalar)
+    pub q_mul_complete: Selector,
+    /// Fixed-base full-width scalar multiplication
+    pub q_mul_fixed: Selector,
+    /// Fixed-base signed short scalar multiplication
+    pub q_mul_fixed_short: Selector,
+    /// Witness point
+    pub q_point: Selector,
+    /// Witness full-width scalar for fixed-base scalar mul
+    pub q_scalar_fixed: Selector,
+    /// Witness signed short scalar for full-width fixed-base scalar mul
+    pub q_scalar_fixed_short: Selector,
+    /// Permutation
+    pub perm: Permutation,
+}
+
+/// A chip implementing EccInstructions
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct EccChip<C: CurveAffine> {
+    config: EccConfig,
+    _marker: PhantomData<C>,
+}
+
+impl<C: CurveAffine> Chip<C::Base> for EccChip<C> {
+    type Config = EccConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+impl<C: CurveAffine> EccChip<C> {
+    pub fn construct(config: <Self as Chip<C::Base>>::Config) -> Self {
+        Self {
+            config,
+            _marker: PhantomData,
+        }
+    }
+
+    #[allow(non_snake_case)]
+    pub fn configure(
+        meta: &mut ConstraintSystem<C::Base>,
+        advices: [Column<Advice>; 10],
+    ) -> <Self as Chip<C::Base>>::Config {
+        let config = EccConfig {
+            advices,
+            lagrange_coeffs: [
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+            ],
+            fixed_z: meta.fixed_column(),
+            q_add_incomplete: meta.selector(),
+            q_add: meta.selector(),
+            q_mul_hi: meta.selector(),
+            q_mul_lo: meta.selector(),
+            q_mul_decompose_var: meta.selector(),
+            q_mul_complete: meta.selector(),
+            q_mul_fixed: meta.selector(),
+            q_mul_fixed_short: meta.selector(),
+            q_point: meta.selector(),
+            q_scalar_fixed: meta.selector(),
+            q_scalar_fixed_short: meta.selector(),
+            perm: Permutation::new(
+                meta,
+                &[
+                    advices[0].into(),
+                    advices[1].into(),
+                    advices[2].into(),
+                    advices[3].into(),
+                    advices[4].into(),
+                    advices[6].into(),
+                    advices[7].into(),
+                    advices[8].into(),
+                    advices[9].into(),
+                ],
```

Comment:
```suggestion
                advices
                    .iter()
                    .map(|a| (*a).into())
                    .collect::<ArrayVec<Column<Any>, 10>>()
                    .as_slice(),
```

Reply:
I realised the ECC chip `configure()` should take in a `perm: Permutation`, now that we're defining a single top-level permutation.